### PR TITLE
change import verifier to check individual packages

### PR DESCRIPTION
**Pull request by deads2k**
This updates our import verifier to check individual packages instead of checking entire sub-trees.  This works better for our APIs, since the import trees under `install` and `validation` aren't going to affect our client.

It also allows multiple related paths and lets you specify individual packages so that people don't accidentally creep under a given tree, making the choice explicit.

### CODE REVIEW

**Comment by stevekuznetsov** [comment]
Err -- you probably want to rebase your changes on top of what I had to change to make the k8s PR work

**Comment by stevekuznetsov** [comment]
And we should probably get that merged first, so we serialize these efforts and don't end up with weird drift

**Comment by stevekuznetsov** [comment]
/assign

@soltysh why you unassign me?

**Comment by liggitt** [comment]
> @soltysh why you unassign me?

he didn't, it's a github bug in how it renders assignments/unassignments

**Comment by stevekuznetsov** [comment]
@kargakis wait am I seeing this right that MY account assigned people up there? I was not in the office 20 minutes ago and that looks like Blunderbuss .... wtf?

**Comment by stevekuznetsov** [comment]
@liggitt oh, ok. I was not aware there were bugs in that. Thanks.

**Comment by deads2k** [comment]
> And we should probably get that merged first, so we serialize these efforts and don't end up with weird drift

We can reconcile during a rebase, but this needs to move ahead now and tighten what we have in order to move ahead with separating a workable client.

**Comment by stevekuznetsov** [comment]
@deads2k dont understand why you want to do rebase reconciling ... seems more painful than necessary

**Comment by deads2k** [comment]
> @deads2k dont understand why you want to do rebase reconciling ... seems more painful than necessary

Because it will give both forks time to mature.  If needed down here we run two and then move the pieces we need.  I definitely don't want to be waiting on upstream pulls to merge to make this do what I need it to do.

**Comment by stevekuznetsov** [comment]
The only reason the upstream pull hasn't yet merged is YOUR review, though.

**Comment by stevekuznetsov** [comment]
I mean, this is code that's supposed to work for you. if you think a rebase reconciliation workflow is better for you, ok. no problems there

File: hack/import-restrictions.json [hunk 1]
```diff
@@ -1,9 +1,228 @@
 [
-   {
-      "baseImportPath" : "pkg/image",
-      "allowedImports" : [
-         "pkg/client",
-         "vendor"
-      ]
-   }
+  {
+    "checkedPackages": [
```
Comment (stevekuznetsov) [inline hack/import-restrictions.json]: This API is a little hard to follow. It's not clear we are implicitly allowing imports of self and of other checked packages

File: hack/import-restrictions.json [hunk 2]
```diff
@@ -1,9 +1,228 @@
...
+      "github.com/openshift/origin/pkg/authorization/apis/authorization",
+      "github.com/openshift/origin/pkg/authorization/apis/authorization/v1"
+    ],
+    "allowedImportPackageRoots": [
+      "vendor/k8s.io/apimachinery",
+      "vendor/github.com/gogo/protobuf"
+    ],
+    "allowedImportPackages": [
+      "vendor/github.com/google/gofuzz",
```
Comment (stevekuznetsov) [inline hack/import-restrictions.json]: Upstream to match what was being written before I stripped vendor from the front of any declarations. IMO it's a little cleaner to look at from a config point of view since you don't really care and there are no nested vendors (right?)

File: tools/import-verifier/import-verifier.go [hunk 3]
```diff
@@ -89,23 +98,58 @@ func (i *ImportRestriction) forbiddenImportsFor(pkg Package) []string {
 //   - of a package under the rootPackage
```
Comment (stevekuznetsov) [inline tools/import-verifier/import-verifier.go]: Need to update this godoc

File: tools/import-verifier/import-verifier.go [hunk 4]
```diff
@@ -89,23 +98,58 @@ func (i *ImportRestriction) forbiddenImportsFor(pkg Package) []string {
...
+		return true
 	}
 
-	return importsBelowRoot && !importsBelowBase && !importsAllowed
+	for _, allowedPackage := range i.AllowedImportPackages {
+		if strings.HasPrefix(allowedPackage, "vendor") {
+			allowedPackage = rootPackage + "/" + allowedPackage
+		}
+		if packageToCheck == allowedPackage {
+			return true
```
Comment (stevekuznetsov) [inline tools/import-verifier/import-verifier.go]: The logic here isn't super clear (and it can't be simpler) so that's why I was using named booleans and delegating logic to keep the level of indentation down. As this is the "meat" of the check IMO we should keep it that way:


```go
func (i *ImportRestriction) isAllowed(packageToCheck string) bool {
    return !i.importsBelowRoot(packageToCheck) || i.isIncludedInRestrictedPackages(packageToCheck) || i.isExplicitlyAllowed(packageToCheck)
}

func (i *ImportRestriction) isExplicitlyAllowed(packageToCheck string) bool {
    return i.importsAllowedPackage(packageToCheck) || i.importsBelowAllowedRoot(packageToCheck)
}
```

File: hack/import-restrictions.json (hunk 2 above)
Comment (deads2k) [inline hack/import-restrictions.json]: > Upstream to match what was being written before I stripped vendor from the front of any declarations. IMO it's a little cleaner to look at from a config point of view since you don't really care and there are no nested vendors (right?)

There are.  See builds

File: tools/import-verifier/import-verifier.go (hunk 4 above)
Comment (deads2k) [inline tools/import-verifier/import-verifier.go]: > The logic here isn't super clear (and it can't be simpler) so that's why I was using named booleans and delegating logic to keep the level of indentation down. As this is the "meat" of the check IMO we should keep it that way:

I couldn't read it.  This splits the early decisions first, then checks the last one.

File: hack/import-restrictions.json (hunk 1 above)
Comment (deads2k) [inline hack/import-restrictions.json]: > This API is a little hard to follow. It's not clear we are implicitly allowing imports of self and of other checked packages

Do you have suggestions?  I don't think I want people to repeat themselves.

**Comment by deads2k** [comment]
This still doesn't have conflicts.  I'm going to kick the verify and tag.

@stevekuznetsov I want to continue the discussion in the threads, I just don't want more dependencies leaking in while we do.

File: hack/import-restrictions.json (hunk 2 above)
Comment (stevekuznetsov) [inline hack/import-restrictions.json]: I meant the other way, the config says "package x is allowed" and imports to X or vendored X are allowed

File: tools/import-verifier/import-verifier.go (hunk 4 above)
Comment (stevekuznetsov) [inline tools/import-verifier/import-verifier.go]: Right, I'm following your flow -- I meant just factor it so the method isnt 30 LOC and 3 levels of indentation deep

**Comment by stevekuznetsov** [comment]
Upstream merged, but do as you will

/lgtm
